hazelcast/src/main/java/com/hazelcast/sql/impl/exec/io/ReceiveSortMergeExec.java

```diff
@@ -75,28 +84,49 @@ public ReceiveSortMergeExec(
         }
 
         sorter = new MergeSort(sources, new SortKeyComparator(ascs));
+
+        if (fetch != null || offset != null) {
+            this.fetch = new Fetch(fetch, offset);
+        } else {
+            this.fetch = null;
+        }
     }
 
     @Override
     protected void setup0(QueryFragmentContext ctx) {
         inbox.setup();
+
+        if (fetch != null) {
+            fetch.setup(ctx);
+        }
     }
 
     @Override
     public IterationResult advance0() {
-        List<Row> rows = sorter.nextBatch();
-        boolean done = sorter.isDone();
+        while (true) {
+            List<Row> rows = sorter.nextBatch();
+            boolean done = sorter.isDone();
 
-        if (rows == null) {
-            curBatch = EmptyRowBatch.INSTANCE;
+            if (rows == null) {
+                curBatch = EmptyRowBatch.INSTANCE;
 
-            return done ? IterationResult.FETCHED_DONE : IterationResult.WAIT;
-        } else {
-            RowBatch batch = new ListRowBatch(rows);
+                return done ? IterationResult.FETCHED_DONE : IterationResult.WAIT;
+            } else {
+                RowBatch batch = new ListRowBatch(rows);
+
+                if (fetch != null) {
+                    batch = fetch.apply(batch);
```

Comment:
We may avoid the copying of rows, if we integrate fetching into the `sorter`.

Reply:
Integrated FETCH/OFFSET into the merge sorter in https://github.com/hazelcast/hazelcast/pull/18091/commits/3cc58e602a8a164202f8e834241d2bc09f055e8f